### PR TITLE
feat(cmd): adding timeout flag

### LINKED ISSUE #3575: improvement(cmd): add timeout flag
Our current RPC client, created for the CLI, does not contain any timeouts. We can extend our basic flags with `--timeout`, allowing the user to specify the maximum duration for a request.

It should be added [here](https://github.com/celestiaorg/celestia-node/blob/79295fc5560a0b14d110e02f6a424446ecc58807/cmd/rpc.go#L23), so non-empty timeout could be used [here](https://github.com/celestiaorg/celestia-node/blob/79295fc5560a0b14d110e02f6a424446ecc58807/cmd/rpc.go#L76)

### PR DESCRIPTION
**Pull request by sebasti810**
fixes #3575

### CODE REVIEW

**Comment by jcstein** [comment]
does this time out the transaction in the mempool, or just in the client?

**Comment by sebasti810** [comment]
just in the client. is it supposed to also set it for the mempool?

**Comment by jcstein** [comment]
I'm thinking of a case where I send transaction with nonce x, it times out, and then when I go to send another transaction, also with nonce x, it errors because it has the same nonce as the tx that is already timed out on client, but in the mempool

**Comment by vgonkivs** [comment]
But node does not support nonce so far. We agreed that's it is not necessary.

**Comment by distractedm1nd** [comment]
> I'm thinking of a case where I send transaction with nonce x, it times out, and then when I go to send another transaction, also with nonce x, it errors because it has the same nonce as the tx that is already timed out on client, but in the mempool

This does not happen anymore - the nonce error was fixed by upgrading app :)

> But node does not support nonce so far. We agreed that's it is not necessary.

I don't think he's saying we need to support nonce, I think @jcstein was just describing that what happens with a nonce in the background was throwing errors to the user, and that a timeout for txs in the mempool could fix this. 

What if we add a timeout to `TxConfig` in a separate issue?

**Comment by vgonkivs** [comment]
I'd rather pass it through the context.

**Comment by distractedm1nd** [comment]
Fair, I guess we can't have a timeout in the transaction itself that gets submitted, right?

**Comment by vgonkivs** [comment]
I don't want to grow the TxConfig. Timeout is smth we can configure through the context.
wdyt @distractedm1nd ?

**Comment by jcstein** [comment]
> This does not happen anymore - the nonce error was fixed by upgrading app :)
😄 🚢

**Comment by renaynay** [comment]
@jcstein Did you test this PR already?

**Review by renaynay — APPROVED** [review]

**Review by cristaloleg — APPROVED** [review]

**Review by walldiss — APPROVED** [review]

File: cmd/rpc.go [hunk 1]
```diff
@@ -37,6 +39,13 @@ func RPCFlags() *flag.FlagSet {
 		"Authorization token",
 	)
 
+	fset.DurationVar(
+		&timeoutFlag,
+		"timeout",
+		0,
+		"Timeout for RPC requests (e.g. 30s, 1m)",
```
Comment (vgonkivs) [inline cmd/rpc.go:46]: Could you describe how the user should interact with the flag? In what units the timeout should be passed, bc duration represents time in nanoseconds.?

**Review by vgonkivs — CHANGES_REQUESTED** [review]
Could you describe how the user should interact with the flag? In what units the timeout should be passed, bc duration represents time in nanoseconds.?

**Comment by sebasti810** [comment]
@vgonkivs The flag docstring prints the usage, e.g. 30s, 1m. This is because pflag has a DurationVar option, which handles this parsing underneath. 

And it's tested, this is the output you get when the timeline is exceeded
image

**Comment by jcstein** [comment]
> 
@jcstein Did you test this PR already?

@renaynay just saw this, no I haven't tested it yet

File: cmd/rpc.go [hunk 2]
```diff
@@ -73,7 +82,12 @@ func InitClient(cmd *cobra.Command, _ []string) error {
 		}
 	}
 
-	client, err := rpc.NewClient(cmd.Context(), requestURL, authTokenFlag)
+	var opts []rpc.RPCClientOption
+	if timeoutFlag > 0 {
+		opts = append(opts, rpc.WithTimeout(timeoutFlag))
+	}
+
+	client, err := rpc.NewClient(cmd.Context(), requestURL, authTokenFlag, opts...)
```
Comment (Wondertan) [inline cmd/rpc.go:90]: There is a simpler way to achieve the same result.

We can pass through the timeout in the context. The timeout is respected by the RPC HTTP client performing the request, which avoids creating a new option for the client.

**Review by Wondertan — CHANGES_REQUESTED** [review]
Hey @sebasti810! 

Thank you for the contribution

**Review by renaynay — DISMISSED** [review]

**Review by Wondertan — DISMISSED** [review]

**Review by cristaloleg — APPROVED** [review]

File: cmd/rpc.go [hunk 3]
```diff
@@ -73,6 +82,12 @@ func InitClient(cmd *cobra.Command, _ []string) error {
 		}
 	}
 
+	if timeoutFlag > 0 {
+		// we don't cancel, because we want to keep this context alive outside the InitClient Function
+		ctx, _ := context.WithTimeout(cmd.Context(), timeoutFlag)
```
Comment (cristaloleg) [inline cmd/rpc.go:87]: Can you add `//nolint:govet` here please?

**Review by cristaloleg — APPROVED** [review]

**Review by walldiss — APPROVED** [review]

**Pull request merged**